### DISCUSSION TITLE
profiling memory?

### DISCUSSION THREAD

**Discussion by jtorjo** [emscripten-core/emscripten#19149] [Q&A]
Hi guys,

I have a (.net webassembly) app that is rather huge. When ran for 1h+, it usually runs out of memory.
So obviously, there are some leaks. But since we can't really do memory profiling (that arraybuffer is just opaque), I'm hitting a dead end.

Right now I'm trying to figure out scenarios that will trigger memory increase, but so far, we've solved most of them, and we're now left with the "slow death of a thousand cuts" leaks -- those that do increase memory, but very very slowly. And inconsistently.

As far as I can see, there's nothing out there to help with this.

What do you guys suggest I do?

Is there a simple way to somehow log all "allocation"s and "deallocation"s and then have a way to trigger a way two snapshots, and compute differences?

**Comment by kripken** [comment]
For leaks, see this article: https://web.dev/webassembly-memory-debugging/

There is a simple memory profiler, `--memoryprofiler`, but it doesn't track allocation sites I think, https://groups.google.com/g/emscripten-discuss/c/HtAzVposlX4?pli=1

There is also a tracing API you can build profiling from: https://emscripten.org/docs/api_reference/trace.h.html
